### ISSUE TITLE
QueuedEmails: Emails not locked while processing

### ISSUE THREAD

**Issue by Manuel-Innovapps** [nopSolutions/nopCommerce#6958]
nopCommerce version: 4.60.x

Steps to reproduce the problem: 
1. Setup a small and slow email server (in our case a on-prem hosted exchange server from one of our customers)
2. Setup the tasks to send emails to run every minute (as per nop defaults, same for the task timeout)
3. Create a new email campaign / newsletter (about 400 recipients in our case, only customers who signed up to the newsletter this year)
4. Send the email campaign

As the email tasks always fetches the first 500 not send emails, in our case, each minute, the task fetched all 400 emails. In our case the total duration to send all emails was about 25 minutes (first to last email send). The tasks scheduler spins up a new task every minute, but does not appear to shut down the previous task correctly. 

This means that in our case we had about 20 tasks running in parallel at some time all processing the same emails (because they are only fetched from the database while the task is booting up). This resulted in multiple of our customers being notified up to 20 times with the same campaign / newsletter email.

This irritated a lot of customers of this store and made the store owner truly unhappy to notify so many customers.

---

We would expect that each email that is picked up in a task is "locked" in some kind, so that a task starting after the current task, knows to skip all those "locked" emails.

And while we are underway we should probably also remove the hardcoded limit of 500 emails that are fetch for each task execution.

**Comment by AndreiMaz** [comment]
@Manuel-Innovapps Could you please check it one more time? We already validate. Please check the "IsTaskAlreadyRunning" method of the \Nop.Services\ScheduleTasks\ScheduleTaskRunner.cs and how it's used

**Comment by Manuel-Innovapps** [comment]
Hey @AndreiMaz 

I was not aware of this piece of code, and indeed it looks like it should be doning it's thing. But that's the conclusion I came up based on the logs. I have attached you the relevant logs from our site. You should mainly use the file without the full message, as excel is incapable of working with multi-line logs :)

[ZS-45_Logs_without_FullMessage.xlsx](https://github.com/nopSolutions/nopCommerce/files/13700211/ZS-45_Logs_without_FullMessage.xlsx)
[ZS-45_Logs_with_FullMessage.xlsx](https://github.com/nopSolutions/nopCommerce/files/13700212/ZS-45_Logs_with_FullMessage.xlsx)

There you can see that the email task is timeouting 10 times at the beginning (which is at the bottom of the file) and one time at the end. In between we receive a bunch of error messages.

The shop is a complete clean installation, without any custom plugins, only with a SevenSpikes Theme and some according plugins. But I would not expect them to modify the task scheduler.

Hope that helps!

**Comment by skoshelev** [comment]
Hi @Manuel-Innovapps. It's hard to tell from your log that something is going wrong. It can be seen that you are configured to send messages after 120 seconds. You can also see that the task started sending messages at 14:43:38 and tried to send 924 letters (and maybe more, since apparently only those that could not be sent were recorded in the log). The last message that was not sent has a time in the log of 14:45:00, that is, the task worked for 1 minute and 22 seconds, and the next task tried to start only at 14:47:11 (two minutes after the last message). If there was a problem that you are talking about, I believe the time of the next entry would be approximately 14:45, but since messages were still being sent at that time, the new task did not start.

I understand that something is going wrong, but understanding what exactly is quite problematic. You can implement additional logging in the QueuedMessagesSendTask itself. Add, for example, a log entry that the task has started, how many emails it will send, and that the task has ended. You can also display information about how many letters were sent successfully and how many were not.

**Comment by Manuel-Innovapps** [comment]
@skoshelev That's the main problem. In our queued emails we only have 391 items that should be send. Were are you getting the 120 seconds from? I'm unable to find where I should have configured that.

I will try to reproduce this locally, once I get time for this!

**Comment by skoshelev** [comment]
@Manuel-Innovapps I assumed 120 seconds from the error interval at the beginning of the log, I could have been wrong here. But it looks like I managed to find the problem. By default, task requests wait only 100 seconds for a response from the server. Since the task takes longer to complete in your case, and on the server, the request apparently has a longer lifespan, the new task starts earlier than necessary. I advise you to increase the request execution waiting time for scheduled tasks. This can be done by setting "Schedule task run timeout" in AppSettings. I would advise you to sync this value with your HTTP server settings

**Comment by skoshelev** [comment]
@Manuel-Innovapps. No, when I talked about synchronizing timeouts, I was talking about the lifetime of a request on an HTTP server, for example, the "connection time out" setting on IIS. 
But I agree, adding a setting for the number of messages at a time would be very useful

**Comment by skoshelev** [comment]
Closed #6958

**Comment by Manuel-Innovapps** [comment]
Hey @skoshelev 

thanks for the clarification and the fix. 50 emails per Minute sounds like a lot more sensible default!
